frontend/utils/distribution/monthly-event-parser.js

```diff
@@ -0,0 +1,199 @@
+    const paymentA = payments[i]
+    for (let j = i + 1; j < payments.length; j++) {
+      const paymentB = payments[j]
+
+      // Were paymentA and paymentB between the same two users?
+      if ((paymentA.from === paymentB.from && paymentA.to === paymentB.to) ||
+        (paymentA.to === paymentB.from && paymentA.from === paymentB.to)) {
+        // Add or subtract paymentB's amount to paymentA's amount, depending on the relative
+        // direction of the two payments:
+        paymentA.amount += (paymentA.from === paymentB.from ? 1 : -1) * paymentB.amount
+
+        // Remove paymentB from payments, and decrement the inner sentinal loop variable:
+        payments = payments.filter((_, paymentIndex) => { return paymentIndex !== j })
+        j--
+      }
+    }
+  }
+  return payments
+}
+
+// DRYing function meant for accumulating late payments from a previous cycle
+function addDistributions (paymentsA: Array<any | Object>, paymentsB: Array<any | Object>): Array<any | {|from: string, to: string, amount: number|}> {
+  return reduceDistribution([paymentsA, paymentsB].flat())
+}
+
+// DRYing function meant for chipping away a cycle's todoPayments distribution using that cycle's completedMonthlyPayments:
+function subtractDistributions (paymentsA: Array<any | Object>, paymentsB: Array<any | Object>): Array<any | {|from: string, to: string, amount: number|}> {
+  // Don't modify any payment list/objects parameters in-place, as this is not intended:
+  paymentsB = JSON.parse(JSON.stringify(paymentsB))
+
+  // Reverse the sign of the second operand's amounts so that the final addition is actually subtraction:
+  paymentsB = paymentsB.map((p) => {
+    p.amount *= -1
+    return p
+  })
+
+  return addDistributions(paymentsA, paymentsB)
+}
+
+// This algorithm is responsible for calculating the monthly-rated distribution of
+// payments. monthlyRated = True if time-weighted monthly, false if purely pro-rated.
+function parseMonthlyDistributionFromEvents (distributionEvents: Array<Object>, minCome: number, monthlyRated: Boolean, adjusted: Boolean): Array<any | {|from: string, to: string, amount: number|}> {
+  distributionEvents = JSON.parse(JSON.stringify(distributionEvents))
+
+  const lastEvent = distributionEvents[distributionEvents.length - 1]
+  const newLastEvent = {
+    type: 'startCycleEvent',
+    data: {
+      cycle: Math.floor(lastEvent.data.cycle) + 1,
+      latePayments: [] // List to be populated later, by the events-parser
+    }
+  }
+  // Add blank 'startCycleEvent's in after the distributionEvents
+  distributionEvents.push(newLastEvent)
+
+  // The following list variable is for DRYing out our calculations of the each cycle's final
+  // income distributions.
+  let groupMembers = []
+
+  // Convenience function for retreiving a user by name:
+  const getUser = function (userName) {
+    for (const member of groupMembers) {
+      if (member.name === userName) {
+        return member
+      }
+    }
+  }
```

Comment:
This looks like it could be written in a single line with `.find()`?

E.g. someting like:

```js
const getUser = name => groupMembers.find( ... )
```

Reply:
Good observation. Fixed.